crates/uv-configuration/src/dev.rs

```diff
@@ -1,445 +1,350 @@
-    pub fn contains(&self, group: &GroupName) -> bool {
-        match self {
-            DevMode::Exclude => false,
-            DevMode::Include | DevMode::Only => group == &*DEV_DEPENDENCIES,
-        }
-    }
-}
-
-#[derive(Default, Debug, Clone)]
-pub struct DevGroupsSpecification {
-    /// Legacy option for `dependency-groups.dev` and `tool.uv.dev-dependencies`.
-    ///
-    /// Requested via the `--dev`, `--no-dev`, and `--only-dev` flags.
-    dev: Option<DevMode>,
+/// Manager of all dependency-group decisions and settings history.
+///
+/// This is an Arc mostly just to avoid size bloat on things that contain these.
+#[derive(Debug, Default, Clone)]
+pub struct DevGroupsSpecification(Arc<DevGroupsSpecificationInner>);
 
-    /// The groups to include.
+/// Manager of all dependency-group decisions and settings history.
+#[derive(Debug, Default, Clone)]
+pub struct DevGroupsSpecificationInner {
+    /// Groups to include.
+    include: IncludeGroups,
+    /// Groups to exclude (always wins over include).
+    exclude: Vec<GroupName>,
+    /// Whether an `--only` flag was passed.
     ///
-    /// Requested via the `--group` and `--only-group` options.
-    groups: Option<GroupsSpecification>,
+    /// If true, users of this API should refrain from doing "non-group things".
+    only_groups: bool,
+    /// The "raw" flags/settings we were passed for diagnostics.
+    history: DevGroupsSpecificationHistory,
 }
 
-#[derive(Debug, Clone)]
-pub enum GroupsSpecification {
-    /// Include dependencies from the specified groups alongside the default groups (omitting
-    /// those default groups that are explicitly excluded).
-    ///
-    /// If the `include` is `IncludeGroups::Some`, it is guaranteed to omit groups in the `exclude`
-    /// list (i.e., they have an empty intersection).
-    Include {
-        include: IncludeGroups,
-        exclude: Vec<GroupName>,
-    },
-    /// Include dependencies from the specified groups, omitting any default groups.
-    ///
-    /// If the list is empty, no group will be included.
-    Explicit { include: Vec<GroupName> },
-    /// Only include dependencies from the specified groups, exclude all other dependencies.
+impl DevGroupsSpecification {
+    /// Create from history.
     ///
-    /// The `include` list is guaranteed to omit groups in the `exclude` list (i.e., they have an
-    /// empty intersection).
-    Only {
-        include: Vec<GroupName>,
-        exclude: Vec<GroupName>,
-    },
-}
-
-impl GroupsSpecification {
-    /// Create a [`GroupsSpecification`] that includes the given group.
-    pub fn from_group(group: GroupName) -> Self {
-        Self::Include {
-            include: IncludeGroups::Some(vec![group]),
-            exclude: Vec::new(),
+    /// This is the "real" constructor, it's basically taking raw CLI flags but in
+    /// a way that's a bit nicer for other constructors to use.
+    pub fn from_history(history: DevGroupsSpecificationHistory) -> Self {
```

Comment:
This probably doesn't need to be `pub` right? It'd be nice to only have one public constructor.

Reply:
NB there's already other convenience constructors but yeah making this non-pub is fine.